library/std/src/sys/thread_local/no_threads.rs

```diff
@@ -70,24 +87,39 @@ impl<T> LazyStorage<T> {
     /// has occurred.
     #[inline]
     pub fn get(&'static self, i: Option<&mut Option<T>>, f: impl FnOnce() -> T) -> *const T {
-        let value = unsafe { &*self.value.get() };
-        match value {
-            Some(v) => v,
-            None => self.initialize(i, f),
+        if self.state.get() == State::Alive {
+            self.value.get() as *const T
+        } else {
+            self.initialize(i, f)
```

Comment:
This will hurt codegen for thread locals that otherwise had option optimization (e.g., if you store NonNull pointer today, this would optimize to a single load + test, rather than loading from two different memory locations).

I'm not sure how much that matters in practice, but it seems unfortunate to pay the cost for all users. Ideally, we'd only do so for the case where alignof(Option.Some.0) != alignof(Option), but I'm not sure if that's possible to check in macros / const-fn to optimize this.

Does this code get used only in the no_threads case? Or does it end up exported for the actually threaded cases too?